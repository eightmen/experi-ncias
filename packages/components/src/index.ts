export * from './Box'
export * from './Flex'
export * from './Grid'
export * from './Button'
export * from './Link'
export * from './Paragraph'
export * from './Text'
export * from './Heading'
export * from './Image'
export * from './Card'
export * from './Label'
export * from './Input'
export * f