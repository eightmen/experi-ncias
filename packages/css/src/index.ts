
import type {
  CSSObject,
  ThemeUIStyleObject,
  ThemeDerivedStyles,
  Theme,
  ThemeUICSSObject,
} from './types'

export * from './types'

/**
 * Allows for nested scales with shorthand values
 * @example
 * {
 *   colors: {
 *     primary: { __default: '#00f', light: '#33f' }
 *   }
 * }
 * css({ color: 'primary' }); // { color: '#00f' }
 * css({ color: 'primary.light' }) // { color: '#33f' }
 */
export const THEME_UI_DEFAULT_KEY = '__default'

const hasDefault = (
  x: unknown
): x is { [THEME_UI_DEFAULT_KEY]: string | number } => {
  return typeof x === 'object' && x !== null && THEME_UI_DEFAULT_KEY in x
}

/**
 * Extracts value under path from a deeply nested object.
 * Used for Themes, variants and Theme UI style objects.
 * Given a path to object with `__default` key, returns the value under that key.
 *
 * @param obj a theme, variant or style object
 * @param path path separated with dots (`.`)
 * @param fallback default value returned if get(obj, path) is not found
 */
export function get(
  obj: object,
  path: string | number | undefined,
  fallback?: unknown,
  p?: number,
  undef?: unknown
): any {
  const pathArray = path && typeof path === 'string' ? path.split('.') : [path]
  for (p = 0; p < pathArray.length; p++) {
    obj = obj ? (obj as any)[pathArray[p]!] : undef
  }

  if (obj === undef) return fallback

  return hasDefault(obj) ? obj[THEME_UI_DEFAULT_KEY] : obj
}

export const getObjectWithVariants = (obj: any, theme: Theme): CSSObject => {
  if (obj && obj['variant']) {
    let result: CSSObject = {}
    for (const key in obj) {
      const x = obj[key]
      if (key === 'variant') {
        const val = typeof x === 'function' ? x(theme) : x
        const variant = getObjectWithVariants(get(theme, val as string), theme)
        result = { ...result, ...variant }
      } else {
        result[key] = x as CSSObject
      }
    }