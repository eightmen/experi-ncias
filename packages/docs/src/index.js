// @ts-check

/** @jsx jsx */
import { jsx } from 'theme-ui'

import Layout from './components/layout'

export { default as Banner } from './components/banner'
export { default as Tiles } from './components/tiles'
export { default as Cards } from './components/cards'
export { default as Note } from './components/note'

/**
 * @param {import("gatsby").WrapPageElementBrowserArgs} props
 */
export const WrapPageElement = ({ element, props }) => (
  <Layout {...props} children={element} />
)

/**
 * @see https://docsearch.algolia.com/
 * @param {import("gatsby").RenderBodyArgs} args
 */
export const setDocSearchComponents = ({
  setHeadComponents,
  setPostBodyComponents,
}) => {
  setHeadComponents([
    <link
      key="algolia-css"
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css"
    />,
  ])
  setPostBodyComponents([
    <script
      key="algolia-cdn"
      type="text/javascript"
      src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"
    ></script>,
    <script
      key="algolia-script"
      type="text/javascript"
      dangerouslySetInnerHTML={{
        __html: `
          var lastSearchInput = null; 
          var observer = new MutationObserver(function () {
            var searchSelector = "#algolia-docs-search";
            var searchInput = document.querySelector(searchSelector);

            if (searchInput && (lastSearchInput !== searchInput)) {
              docsearch({
                apiKey: "84ed820927eee5fa5018c9f1abe70390",
                indexName: "theme-ui",
                inputSelector: searchSelector,
                debug: true,
                transformData: function(hits) {
                  for (const hit of hits) {
                    hit.url = hit.url.replace(
                      "https://theme-ui.com",
                      window.location.origin
                    )
                  }
                }
              })
              lastSearchInput = searchInput;
            }
          });

          document.addEventListener("DOMContentLoaded", function() {
            observer.observe(document, {
              childList: true,
              subtree: true
            });
          });        
      `,
      }}
    />,
  ])
}
