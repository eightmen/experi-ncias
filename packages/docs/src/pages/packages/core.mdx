---
title: '@theme-ui/core'
---

import Readme from '@theme-ui/core/README.md'

<Readme />
