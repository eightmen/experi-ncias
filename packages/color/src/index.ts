import * as P from 'polished'
import { ColorModesScale, get, Theme } from '@theme-ui/css'

type Color = ColorModesScale[string]

/**
 * Get color from theme.colors
 */
export const getColor = (theme: Theme, color: Color) => {
  if (typeof color === 'object') {
    color = Array.isArray(color)
      ? color[0]
      : (color as Exclude<typeof color, string & {}>).__default
  }

  if (process.env.NODE_ENV !== 'production') {
    if (color && /^var\(--theme-ui-colors-(.+)\)$/.test(color)) {
      throw new Error(
        'A CSS property was passed to `getColor`. ' +
          '`theme.colors` contains CSS custom properties. ' +
          'Use `theme.rawColors` instead.'
      )
    }
  }

  return get(
    theme,
    'rawColors' in theme ? `rawColors.${color}` : `colors.${color}`,
    color
  )
}

/**
 * Darken a color by an amount 0–1
 */
export const darken = (c: Color, n: number) => (t: Theme) =>
  P.darken(n, getColor(t, c))
/**
 * Lighten a color by an amount 0–1
 */
export const lighten = (c: Color, n: number) => (t: Theme) =>
  P.lighten(n, getColor(t, c))
/**
 * Rotate the hue of a color by an amount 0–360
 */
export const rotate = (c: Color, d: number) => (t: Theme) =>
  P.adjustHue(d, getColor(t, c))

/**
 * Set the hue of a color to a degree 0–360
 */
export const hue = (c: Color, h: number) => (t: Theme) =>
  P.setHue(h, getColor(t, c))
/**
 * Set the saturation of a color to an amount 0–1
 */
export const saturation = (c: Color, s: number) => (t: Theme) =>
  P.setSaturation(s, getColor(t, c))
/**
 * Set the lightness of a color to an amount 0–1
 */
export const lightness = (c: Color, l: number) => (t: Theme) =>
  P.setLightness(l, getColor(t, c))
/**
 * Desaturate a color by an amount 0–1
 */
export const desaturate = (c: Color, n: number) => (t: Theme) =>
  P.desaturate(n, getColor(t, c))
/**
 * Saturate a color by an amount 0–1
 */
export const saturate = (c: Color, n: number) => (t: Theme) =>
  P.saturate(n, getColor(t, c))

/**
 * Shade a color by an amount 0–1
 */
export const shade = (c: Color, n: number) => (t: Theme) =>
  P.shade(n, getColor(t, c))
/**
 * Tint a color by an amount 0–1
 */
export const tint = (c: Color, n: number) => (t: Theme) =>
  P.tint(n, getColor(t, c))

export const transparentize = (c: Color, n: number) => (t: Theme) =>
  P.transparentize(n, getColor(t, c))
/**
 * Set the transparency of a color to an amount 0-1
 */
export const alpha = (c: Color, n: number) => (t: Theme) =>
  P.rgba(getColor(t, c), n)

/**
 * Mix two colors by a specific ratio
 */
export const mix =
  (a: Color, b: Color, n = 0.5) =>
  (t: Theme) =>
    P.mix(n, getColor(t, a), getColor(t, b))

/**
 * Get the complement of a color
 */
export const complement = (c: Color) => (t: Theme) =>
  P.complement(getColor(t, c))

/**
 * Get the inverted color
 */
export const invert = (c: Color) => (t: Theme) => P.invert(getColor(t, c))

/**
 * Desaturate the color to grayscale
 */
export const grayscale = (c: Color) => desaturate(c, 1)
