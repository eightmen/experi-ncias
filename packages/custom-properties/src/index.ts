import pluralize from 'pluralize'
import { Theme } from '@theme-ui/css'

// Simplified validator based on spec https://www.w3.org/TR/CSS22/syndata.html#value-def-identifier
// Does not check for "cannot start with a digit, two hyphens, or a hyphen followed by a digit"
const keyValidator = /^[A-z0-9][\w-]*$/

interface CustomProperties {
  [key: string]: string | number
}

export default function makeCustomProperties(theme: Theme, prefix?: string) {
  const customProperties: CustomProperties = {}

  const generateProperties = (object: object, previousKey?: string) => {
    Object.entries(object).forEach(([key, value]) => {
      let formattedKey = pluralize(key, 1) || key

      if (
        process.env.NODE_ENV !== 'production' &&
        !keyValidator.test(formattedKey)
      ) {
        console.warn(
          `[theme-ui] Theme key "${value}" found will produce an invalid CSS custom property. ` +
            'Keys must only contain the following: A-Z, a-z, 0-9, hyphen, underscore.'
        )
      }

      if (prefix && !previousKey) {
        formattedKey = `${prefix}-${formattedKey}`