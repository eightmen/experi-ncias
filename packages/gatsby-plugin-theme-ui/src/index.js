import { makeTheme } from '@theme-ui/css/utils'

// add colors, fonts, fontSizes, space, etc
const theme = makeTheme({})

export default theme
