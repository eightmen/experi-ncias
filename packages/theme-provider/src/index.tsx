import React from 'react'
import {
  jsx,
  useThemeUI,
  ThemeProvider as CoreProvider,
  ThemeProviderProps as CoreThemeProviderProps,
  __themeUiDefaultContextValue,
} from '@theme-ui/core'
import { css, Theme } from '@theme-ui/css'
import { ColorModeProvider } from '@theme-ui/color-modes'
import { Global } from '@emotion/react'

const RootStyles = () =>
  jsx(Global, {
    styles: (emotionTheme: unknown) => {
      const theme = emotionTheme as Theme
      const { useRootStyles } = theme.config || theme

      if (useRootStyles === false || (theme.styles && !theme.styles.root)) {
        return null
      }

      const