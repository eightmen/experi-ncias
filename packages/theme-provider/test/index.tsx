
/**
 * @jest-environment jsdom
 * @jsx jsx
 */

import { jsx, useThemeUI, __ThemeUIContext } from '@theme-ui/core'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { matchers } from '@emotion/jest'
import { renderJSON } from '@theme-ui/test-utils'

import { ThemeProvider } from '../src'

expect.extend(matchers)

afterEach(cleanup)

test('renders', () => {
  const json = renderJSON(
    <ThemeProvider theme={{}}>
      <h1>Hello</h1>
    </ThemeProvider>
  )
  expect(json).toMatchSnapshot()
})

test('renders with theme', () => {
  const json = renderJSON(
    <ThemeProvider
      theme={{
        config: {
          useCustomProperties: false,
        },
        colors: {
          primary: 'tomato',
          background: 'white',
          text: 'black',
        },
      }}
    >
      <h1 sx={{ color: 'primary' }}>Hello</h1>
    </ThemeProvider>
  )
  expect(json).toHaveStyleRule('color', 'tomato')
})

test('renders global styles', () => {
  const root = render(
    <ThemeProvider
      theme={{
        fonts: {
          body: 'Georgia,serif',
        },
        lineHeights: {
          body: 1.5,
        },
        fontWeights: {
          body: 500,
        },
        styles: {
          root: {
            fontFamily: 'body',
            fontWeight: 'body',
            lineHeight: 'body',
          },
        },
      }}
    >
      <h1>Hello</h1>
    </ThemeProvider>
  )

  const style = window.getComputedStyle(root.baseElement.parentElement!)
  expect(style.fontFamily).toBe('Georgia,serif')
  expect(style.fontWeight).toBe('500')
  expect(style.lineHeight).toBe('1.5')
})

test('resets body margin', () => {
  const root = render(
    <ThemeProvider theme={{}}>
      <h1>Hello</h1>
    </ThemeProvider>
  )
  const style = window.getComputedStyle(root.baseElement)
  expect(style.margin).toBe('0px')
})

test('does not render invalid global styles', () => {
  const root = render(
    <ThemeProvider theme={{}}>
      <h1>Hello</h1>
    </ThemeProvider>
  )
  const style = window.getComputedStyle(root.baseElement.parentElement!)
  expect(style.fontFamily).toBe('')
  expect(style.fontWeight).toBe('')
  expect(style.lineHeight).toBe('')
})

test('does not render global styles', () => {
  const root = render(
    <ThemeProvider
      theme={{
        config: {
          useRootStyles: false,
        },
        fonts: {
          body: 'Georgia,serif',
        },
        lineHeights: {
          body: 1.5,
        },
        fontWeights: {
          body: 500,
        },
        styles: {
          root: {
            fontFamily: 'body',
          },
        },
      }}
    >
      <h1>Hello</h1>
    </ThemeProvider>
  )
  const style = window.getComputedStyle(root.baseElement)
  expect(style.fontFamily).toBe('')
  expect(style.fontWeight).toBe('')
  expect(style.lineHeight).toBe('')
})

test('adds box-sizing: border-box', () => {
  const root = render(
    <ThemeProvider theme={{}}>
      <h1>Hello</h1>
    </ThemeProvider>
  )
  const style = window.getComputedStyle(root.baseElement)
  expect(style.boxSizing).toBe('border-box')
})

test('does not add box-sizing: border-box', () => {
  const styles: HTMLStyleElement[] = [].slice.call(
    document.querySelectorAll('style')
  )
  styles.forEach((style) => (style.innerHTML = ''))
  const root = render(
    <ThemeProvider
      theme={{
        config: {
          useBorderBox: false,
        },
      }}
    >
      <h1>Hello</h1>
    </ThemeProvider>
  )
  const style = window.getComputedStyle(root.baseElement)
  expect(style.boxSizing).toBe('')
})

test('updates CSS Custom Properties on root element', async () => {
  const DarkModeButton = () => {
    const { colorMode, setColorMode } = useThemeUI()

    if (colorMode === 'dark') return null

    return <button onClick={() => setColorMode!('dark')}>Dark Mode</button>
  }

  const root = render(
    <ThemeProvider
      theme={{
        config: {
          // useCustomProperties defaults to `true`
        },
        colors: {
          text: '#000',
          modes: {
            dark: { text: '#fff' },
          },
        },
      }}
    >
      <DarkModeButton />
    </ThemeProvider>
  )

  const html = root.baseElement.parentElement!

  expect(
    window.getComputedStyle(html).getPropertyValue('--theme-ui-colors-text')
  ).toBe('#000')

  fireEvent.click(root.getByText('Dark Mode'))

  expect(
    window.getComputedStyle(html).getPropertyValue('--theme-ui-colors-text')
  ).toBe('#fff')
})