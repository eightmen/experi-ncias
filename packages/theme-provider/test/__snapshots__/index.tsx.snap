// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<h1>
  Hello
</h1>
`;
